### PR TITLE
KAFKA-5873: add materialized overloads to StreamsBuilder

### PR DESCRIPTION
**Pull request by dguy**
Add overloads for `table` and `globalTable` that use `Materialized`

### CODE REVIEW

**Comment by dguy** [comment]
@guozhangwang, @bbejeck, @mjsax

File: streams/examples/src/main/java/org/apache/kafka/streams/examples/pageview/PageViewTypedDemo.java [hunk 1]
```diff
@@ -145,8 +146,8 @@ public static void main(String[] args) throws Exception {
 
         KStream<String, PageView> views = builder.stream("streams-pageview-input", Consumed.with(Serdes.String(), pageViewSerde));
 
-        KTable<String, UserProfile> users = builder.table(Serdes.String(), userProfileSerde,
-            "streams-userprofile-input", "streams-userprofile-store-name");
+        final Consumed<String, UserProfile> consumed = Consumed.with(Serdes.String(), userProfileSerde);
```
Comment (mjsax) [inline streams/examples/src/main/java/org/apache/kafka/streams/examples/pageview/PageViewTypedDemo.java]: Nit: I would remove var `consumed` -- it's is used just once.

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java [hunk 2]
```diff
@@ -277,450 +251,20 @@
...
-                                                              final String queryableStoreName) {
-        return internalStreamsBuilder.globalTable(topic, new ConsumedInternal<K, V>(), queryableStoreName);
+                                            new ConsumedInternal<>(consumed),
+                                            new MaterializedInternal<>(
+                                                    Materialized.<K, V, KeyValueStore<Bytes, byte[]>>as(
+                                                            internalStreamsBuilder.newStoreName(topic))
+                                                    .withKeySerde(consumed.keySerde)
+                                                    .withValueSerde(consumed.valueSerde),
+                                                    false));
     }
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:281]: If I got this correctly, we have 3 overloads (topic, topic+consumed, topic+ consumed+materialized). Does a 4th with topic+materialized not make sense?

Btw: with consumer+materialized, users can specify key/value serde twice (ie, if default serde must be overwritten, they need to do it twice? -- maybe we can improve this?)

If double specification is required, we should have a check that both key/value serdes are the same?

Applies to `#globalKTable()`, too.

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java [hunk 3]
```diff
@@ -199,46 +201,18 @@
      * query the value of the key on a parallel running instance of your Kafka Streams application.
      *
      * @param topic              the topic name; cannot be {@code null}
-     * @param queryableStoreName the state store name; if {@code null} this is the equivalent of {@link #table(String)}
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:202]: JavaDoc above needs update as `queryableStoreName` gets removed.

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java [hunk 4]
```diff
@@ -827,70 +335,33 @@
...
      * Adds a state store to the underlying {@link Topology}.
      *
-     * @param supplier the supplier used to obtain this state store {@link StateStore} instance
+     * @param builder the builder used to obtain this state store {@link StateStore} instance
      * @param processorNames the names of the processors that should be able to access the provided store
      * @return itself
      * @throws TopologyException if state store supplier is already added
      */
-    public synchronized StreamsBuilder addStateStore(final StateStoreSupplier supplier,
+    public synchronized StreamsBuilder addStateStore(final StoreBuilder builder,
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: We did have a discussion about naming. Didn't we agree to rename `StoreBuilder` to `StateStoreBuilder` ? Might not be part of this PR, but just as a reminder. Or is my memory wrong?

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/MaterializedInternal.java [hunk 5]
```diff
@@ -25,8 +25,16 @@
 
 public class MaterializedInternal<K, V, S extends StateStore> extends Materialized<K, V, S> {
```
Comment (mjsax) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/MaterializedInternal.java]: can be package private I think -- also constructors and all other methods.

**Review by mjsax — COMMENTED** [review]
Couple of comments.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java [hunk 6]
```diff
@@ -113,7 +102,37 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
...
-            consumed.keySerde(), consumed.valueSerde(), Collections.singleton(source), storeSupplier.name(), isQueryable);
+            consumed.keySerde(), consumed.valueSerde(), Collections.singleton(source), storeBuilder.name(), isQueryable);
+
+        internalTopologyBuilder.addStateStore(storeBuilder, name);
+        internalTopologyBuilder.connectSourceStoreAndTopic(storeBuilder.name(), topic);
+
+        return kTable;
+    }
+
+    private <K, V> KTable<K, V> doTable(final ConsumedInternal<K, V> consumed,
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java]: Could we refactor this two private functions a bit differently? What I had in mind is that we can have one private function that include 1) addSource, 2) addProcessor, 3) new KTableImpl, and in the two public `table` APIs we do 0) create processorSupplier, and call the private function, and then 4) addStateStore, 5) connectSource.

In this way the really common path 1/2/3) is consolidated in this private function.

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java [hunk 7]
```diff
@@ -192,15 +196,23 @@ public synchronized void addGlobalStore(final StateStoreSupplier<KeyValueStore>
             valueDeserializer, topic, processorName, stateUpdateSupplier);
     }
 
-    public synchronized void addGlobalStore(final StateStoreSupplier<KeyValueStore> storeSupplier,
+    public synchronized void addGlobalStore(final StoreBuilder<KeyValueStore> storeBuilder,
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java]: This comment is for line 188 above: this function is not used anymore.

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java [hunk 8]
```diff
@@ -199,46 +201,18 @@
...
-     * @param topic         the topic name; cannot be {@code null}
-     * @param storeSupplier user defined state store supplier; cannot be {@code null}
-     * @return a {@link KTable} for the specified topic
-     */
-    public synchronized <K, V> KTable<K, V> table(final String topic,
-                                                  final StateStoreSupplier<KeyValueStore> storeSupplier) {
-        return internalStreamsBuilder.table(null, null, null, null, topic, storeSupplier);
+                                                  final Consumed<K, V> consumed,
+                                                  final Materialized<K, V, KeyValueStore<Bytes, byte[]>> materialized) {
+        Objects.requireNonNull(consumed, "consumed can't be null");
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: This is not introduced in this PR: seems we do not check topic name is `null` anywhere for both `stream` and `table/globalTable`, is this intentional?

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: Also a quick question: if `Consumed` does not specify the same serde as `Materialized`, should we just use different serdes then? I'm asking this mainly because today we will do a deser reading from Kafka and then a ser writing to state store, and maybe we can avoid this deser/ser together as an optimization. But if we allow different serdes here we cannot do that.

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java [hunk 9]
```diff
@@ -907,62 +378,29 @@ public synchronized StreamsBuilder addStateStore(final StateStoreSupplier suppli
...
      * @param processorName         the name of the {@link ProcessorSupplier}
      * @param stateUpdateSupplier   the instance of {@link ProcessorSupplier}
      * @return itself
      * @throws TopologyException if the processor of state is already registered
      */
-    public synchronized StreamsBuilder addGlobalStore(final StateStoreSupplier<KeyValueStore> storeSupplier,
-                                                      final String sourceName,
-                                                      final Deserializer keyDeserializer,
-                                                      final Deserializer valueDeserializer,
+    public synchronized StreamsBuilder addGlobalStore(final StoreBuilder storeBuilder,
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: Hmm.. as a second thought: for this function and `addStateStore` above, should these two functions still be part of `StreamsBuilder`? I know in the old APIs they are in both DSL and PAPI, but that is mainly because the KStreamBuilder is extending TopologyBuilder, so it is easy to get the `processorName`. But now in the `StreamsBuilder` you cannot get the `processorName` any more unless you build the `Topology` and query the specific processor name.

So I'd propose an small addendum to the KIP. Here are a few thoughts I had:

1. Keep these two functions but only with one parameter `StoreBuilder`, and then in `KStream` when `process/transform` is called use `connectProcessorAndStateStores` to connect the stores.

2. Get rid of these two functions, and document that for `process/transform`, you need to call `builder.build().addStateStore()` at the end with the name that you reference inside the `Processor / Transformer`.

3. In `KStream`, deprecate `process(processorSupplier, storeNames...)` with `process(processorSupplier, StoreBuilder...)`, and remove these two functions in `StreamsBuilder`.

WDYT? @mjsax @dguy @bbejeck

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java [hunk 10]
```diff
@@ -123,47 +142,32 @@ public InternalStreamsBuilder(final InternalTopologyBuilder internalTopologyBuil
...
+        internalTopologyBuilder.addGlobalStore(storeBuilder,
                                                sourceName,
                                                consumed.timestampExtractor(),
                                                keyDeserializer,
                                                valueDeserializer,
                                                topic,
                                                processorName,
                                                tableSource);
-        return new GlobalKTableImpl(new KTableSourceValueGetterSupplier<>(storeSupplier.name()));
+        return new GlobalKTableImpl<>(new KTableSourceValueGetterSupplier<K, V>(storeBuilder.name()));
```
Comment (guozhangwang) [inline streams/src/main/java/org/apache/kafka/streams/kstream/internals/InternalStreamsBuilder.java:159]: Nice catch get riding of unchecked warnings!

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsMetadataStateTest.java [hunk 11]
```diff
@@ -88,7 +92,9 @@ public Object apply(final Object value) {
             }
         });
 
-        builder.globalTable("global-topic", "global-table");
+        builder.globalTable("global-topic",
+                            Consumed.with(null, null),
```
Comment (guozhangwang) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsMetadataStateTest.java:96]: This makes me thinking... should we have overloads for 1) topic, 2) topic+consumed, 3) topic+materialized and 4) topic+consumed+materialized ?

Comment (guozhangwang) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsMetadataStateTest.java:96]: EDIT: or maybe we can allow users to specify `null` only for 4)? In that case we can keep not adding 3).

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 9 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: I agree on the issue. Nice catch!!

Side remark:  For `addGlobalStore()` we can keep as is, because `processorName` in not the name of an existing processor, but the name of the newly created processor that maintains the store. Thus, there is no issue with `addGlobalStore()` IMHO, as we don't need to pass in an unknown name of another processor.

Option (2) is not very user friendly IMHO and I have a strong opinion against it.

Second side remark: `connectProcessorAndStateStores()` is not available in the DSL, forcing users to call `builder.build().connectProcessorAndStateStores()`, thus, this is not an option IMHO. It's also not required as one can connect a store to a processor when adding the processor via `process(processorSupplier, storeNames...)`

If we want to stay close to old API, option (1) removing the second parameter from `addStateStore()` makes sense. The store name can be accessed easily via `StoreBuilder#name()` and thus stores can be connected via `process(processorSupplier, storeNames...)`. This is my personally favored option.

Option (3) adding `process(processorSupplier, StoreBuilder...)` might be a syntactic sugar to avoid the `addStateStore()` call. However, we would need to make sure, that two processor can be connected to the same store -- if we deprecate the `process(processorSupplier, storeNames...)` we must match stores via `StoreBuilder` (ie, object identity) -- ie, a user needs to pass in the same `StoreBuilder` to each call of `process(processorSupplier, StoreBuilder...)` to connect all processor to the same store. This, seems to be rather unintuitive and error prone to me... (eg, people try to pass in two different `StoreBuilder` that build equal stores and return the same `storeName` in order to connect two processor to one store -- but to us, it's two stores with a naming conflict and we should throw a `TopologyException` -- this might be rather confusing to users). It also moves away from PAPI usage pattern and I think we should not do that for this case.

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 4 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: It was the other way around. We renamed from `StateStoreBuilder` to `StoreBuilder` as we also added `StoreSupplier`

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 2 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:281]: As for the 4th overload, possibly does make sense.
The double key/value serde thing is a bit of a pain and yes we probably should force them to be the same. Maybe force them into being the same by putting a comment in the docs, something like:
```
// the serdes specified by {@link Consumed} will be used to materialize the store, i.e., serdes in `Materialized` will be ignored.
```
And then we overwrite what is set in `Materialized` with those from `Consumed`. That way the user doesn't need to specify the serdes twice. WDYT? @guozhangwang @bbejeck

File: streams/src/main/java/org/apache/kafka/streams/kstream/internals/MaterializedInternal.java (hunk 5 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/kstream/internals/MaterializedInternal.java]: no it can't be as it is used from `o.a.k.streams`

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 9 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: I agree with @mjsax - i'll update accordingly

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 8 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: I think we should force them to be the same, see comment above. I'll update

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsMetadataStateTest.java (hunk 11 above)
Comment (dguy) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/StreamsMetadataStateTest.java:96]: I think maybe we add the 4th overload

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 9 above)
Comment (dguy) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: I'm wondering if we can actually remove `addGlobalStore` completely from `StreamsBuilder`? It is not needed by the DSL and if users where using the DSL and wanted to manually add a global store they could always use `StreamsBuilder#build` to get a `Topology` and then call `addGlobalStore` on it - WDYT?

Edit: I'm going to remove `addGlobalStore` - i can always put it back...

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 2 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:281]: 4th overload makes sense to me as well. 

I like the idea of placing the serdes in `Consumed` into `Materialized`, but I'm trying to think would there ever be a case where they need to be different? I can't ATM, so I it's a yes for me.

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 9 above)
Comment (bbejeck) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: For option 2 seems a little awkward IMHO, I second what @mjsax says.

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 2 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java:281]: I guess, putting a comment should be fine.

For the actual logic. Not sure if `Materialized` should overwrite `Consumed` -- I would say, if key/value serde is not set for both of them, use defaults from config. If key/value is set in either one of them, use it for both. If both are set, check if they are the same and raise an error if not.

File: streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java (hunk 9 above)
Comment (mjsax) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: I think people want to use a global store in DSL, too. And forcing people to call `build()` to add one, is not a good idea IMHO. (cf option (2)).

Also, if you argue like this, we could remove `addStore`, too, because people can also add the store via `builder.builder().addStore()` -- however, in KIP-120 discussion, it was explicitly requested to add both methods to `StreamsBuilder` to avoid this pattern.

Also note, people who are new and want to use a stateful process() will always ask: how can I add a store? There is not API on `StreamsBuilder`.

Comment (guozhangwang) [reply streams/src/main/java/org/apache/kafka/streams/StreamsBuilder.java]: I think in DSL, users may still wants to access a global state store in `process/transform`, that was the motivation for adding this API.

Personally I'd vote for option 2) above, to keep the APIs succinct without semi-duplicated calls. But seems you all prefer option 1) in terms of user convenience, so I'm fine with that as well.

**Comment by dguy** [comment]
@guozhangwang

**Comment by guozhangwang** [comment]
The latest update LGTM, assuming we are keeping the addStateStore and addGlobalStore in StreamsBuilder. We are currently pondering on whether to still keep them in KIP-182, so a follow-up PR may be coming.

**Comment by dguy** [comment]
merged to trunk
